File: quickstart/output.py
"""Output engine for the quickstart CLI.

Controls verbosity, emoji prefixes, and color theming. All output
flows through this module so a single global flag controls what
the user sees.

Verbosity levels:
  - default (verbose=False): milestones + errors + confirm only
  - verbose (verbose=True): everything (milestones + steps + $ commands)
"""

from __future__ import annotations

import sys

# ── ANSI color palette ──────────────────────────────────────────────
_RESET = "\033[0m"
_BOLD = "\033[1m"
_DIM = "\033[2m"
_RED = "\033[31m"
_GREEN = "\033[32m"
_CYAN = "\033[36m"
_BLUE = "\033[34m"
_MAGENTA = "\033[35m"
_BRIGHT_CYAN = "\033[96m"
_BRIGHT_YELLOW = "\033[93m"
_BRIGHT_MAGENTA = "\033[95m"

_ACCENT_CYCLE: list[str] = [
    _BRIGHT_CYAN,
    _MAGENTA,
    _BLUE,
    _BRIGHT_YELLOW,
    _BRIGHT_MAGENTA,
    _CYAN,
]

# ── Global state ────────────────────────────────────────────────────
_verbose: bool = False
_accent_index: int = 0


def set_verbose(verbose: bool) -> None:
    """Set global verbosity. Called once from __main__.py."""
    global _verbose
    _verbose = verbose


def is_verbose() -> bool:
    return _verbose


def _next_accent() -> str:
    global _accent_index
    color = _ACCENT_CYCLE[_accent_index % len(_ACCENT_CYCLE)]
    _accent_index += 1
    return color


# ── Public output functions ─────────────────────────────────────────

def milestone(msg: str, *, emoji: str = "\u2699\ufe0f ") -> None:
    """Major pipeline milestone. Always visible."""
    accent = _next_accent()
    print(f"\n{emoji}{_BOLD}{accent}{msg}{_RESET}")


def step(msg: str) -> None:
    """Intermediate step. Suppressed unless verbose."""
    if not _verbose:
        return
    print(f"\n{_BOLD}{_CYAN}=> {msg}{_RESET}")


def success(msg: str) -> None:
    """Success/completion. Always visible."""
    print(f"\u2705 {_BOLD}{_GREEN}{msg}{_RESET}")


def error(msg: str) -> None:
    """Error to stderr. Always visible."""
    print(f"\u274c {_BOLD}{_RED}{msg}{_RESET}", file=sys.stderr)


def detail(msg: str) -> None:
    """Detail line. Suppressed unless verbose."""
    if not _verbose:
        return
    print(f"  {_DIM}{msg}{_RESET}")


def info(msg: str) -> None:
    """Informational message. Always visible."""
    print(f"  {msg}")


def confirm(prompt: str) -> bool:
    """Interactive yes/no. Always visible."""
    try:
        answer = input(f"{_BOLD}{prompt}{_RESET}").strip().lower()
        return answer in ("y", "yes")
    except (EOFError, KeyboardInterrupt):
        print()
        return False

File: supervictor/cli/src/output.rs
use std::io::{self, BufRead, Write};
use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};

// ANSI escape codes
const BOLD: &str = "\x1b[1m";
const CYAN: &str = "\x1b[36m";
const GREEN: &str = "\x1b[32m";
const RED: &str = "\x1b[31m";
const YELLOW: &str = "\x1b[33m";
const MAGENTA: &str = "\x1b[35m";
const RESET: &str = "\x1b[0m";

const ACCENT_COLORS: &[&str] = &[CYAN, YELLOW, MAGENTA, GREEN];

static VERBOSE: AtomicBool = AtomicBool::new(false);
static ACCENT_INDEX: AtomicUsize = AtomicUsize::new(0);

/// Enable or disable verbose output globally.
pub fn set_verbose(v: bool) {
    VERBOSE.store(v, Ordering::Relaxed);
}

/// Returns `true` if verbose output is enabled.
pub fn is_verbose() -> bool {
    VERBOSE.load(Ordering::Relaxed)
}

fn next_accent() -> &'static str {
    let idx = ACCENT_INDEX.fetch_add(1, Ordering::Relaxed);
    ACCENT_COLORS[idx % ACCENT_COLORS.len()]
}

/// Major milestone — always visible.
pub fn milestone(msg: &str) {
    let accent = next_accent();
    println!("\n{BOLD}{accent}== {msg} =={RESET}");
}

/// Step header — verbose only.
pub fn step(msg: &str) {
    if !is_verbose() {
        return;
    }
    println!("\n{BOLD}{CYAN}=> {msg}{RESET}");
}

/// Success message.
pub fn success(msg: &str) {
    println!("{GREEN}{msg}{RESET}");
}

/// Error message to stderr.
pub fn error(msg: &str) {
    eprintln!("{RED}{msg}{RESET}");
}

/// Info — always visible, no color.
pub fn info(msg: &str) {
    println!("{msg}");
}

/// Interactive yes/no confirmation. Returns true for "y" or "yes".
pub fn confirm(prompt: &str) -> bool {
    print!("{BOLD}{prompt}{RESET}");
    if io::stdout().flush().is_err() {
        return false;
    }
    let stdin = io::stdin();
    let mut line = String::new();
    match stdin.lock().read_line(&mut line) {
        Ok(0) => false, // EOF
        Ok(_) => {
            let answer = line.trim().to_lowercase();
            answer == "y" || answer == "yes"
        }
        Err(_) => false,
    }
}

/// Reset verbose flag and accent color index (used in tests).
pub fn reset_state() {
    VERBOSE.store(false, Ordering::Relaxed);
    ACCENT_INDEX.store(0, Ordering::Relaxed);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_verbose_default_false() {
        reset_state();
        assert!(!is_verbose());
    }

    #[test]
    fn test_set_verbose() {
        reset_state();
        set_verbose(true);
        assert!(is_verbose());
        set_verbose(false);
        assert!(!is_verbose());
    }

    #[test]
    fn test_accent_cycle() {
        reset_state();
        assert_eq!(next_accent(), CYAN);
        assert_eq!(next_accent(), YELLOW);
        assert_eq!(next_accent(), MAGENTA);
        assert_eq!(next_accent(), GREEN);
        // Wraps around
        assert_eq!(next_accent(), CYAN);
    }
}
